Stark P1 spectroscopy experiment

nkanazawa1989 opened the pull request:
### Summary

This PR adds `StarkP1Spectroscopy` experiment and a dedicated analysis. User can set amplitudes to scan to run this experiment, and the analysis class can convert the amplitudes into frequencies if Stark coefficients are provided. These coefficients can be calibrated with `StarkRamseyXYAmpScan` experiment introduced in #1009 .

### Details and comments

The test experiment result is available in [6799e7ae-414f-4816-b887-b4c3ba498624](https://quantum-computing.ibm.com/experiments/6799e7ae-414f-4816-b887-b4c3ba498624). This experiment result was obtained with the following experiment code:

```python
from qiskit_experiments.library import StarkP1Spectroscopy

exp = StarkP1Spectroscopy((0, ), backend)
exp_data = exp.run().block_for_results()
```

The Stark coefficients can be directly set in the analysis options. By the default setting, the analysis class searches for the coefficients in the experiment service. If analysis results for all coefficients are found (i.e. previously saved in the service), it automatically converts the amplitudes into frequencies for visualization.

A public class method `StarkP1SpectAnalysis.retrieve_coefficients_from_service` is also offered so that a user can set a coefficients dictionary in advance. This is convenient when the experiment instance is run repeatedly, because retrieving analysis data from the service causes a communication overhead. For example,

```python
from qiskit_experiments.library.characterization.analysis import StarkP1SpectAnalysis

# overhead happens only once
coeffs = StarkP1SpectAnalysis.retrieve_coefficients_from_service(
    service=service,
    qubit=0,
    backend=backend.name,
)
exp.analysis.set_options(stark_coefficients=coeffs)

for _ in range(10):
    exp_data = exp.run().block_for_result()
    exp_data.save()
```

User can make a subclass of this analysis class `StarkP1SpectAnalysis` to perform custom analysis. This built-in class doesn't perform any analysis except for visualization. Instead, this class provides a convenient hook `._run_spect_analysis` that takes (x, y, y_err) data and returns a list of `AnalysisResultData`.

nkanazawa1989 commented on qiskit_experiments/library/stark/p1_spectroscopy.py:
```diff
@@ -0,0 +1,283 @@
+              q: ┤ X ├┤ Stark(stark_amp) ├┤M├
+                 └───┘└──────────────────┘└╥┘
+            c: 1/══════════════════════════╩═
+                                           0
```
There is no real delay in the experiment circuit. Stark gate represents a pulsed Stark drive that drifts qubit frequency during the drive.

yaelbh replied:
There's the T1 delay

yaelbh reviewed (commented):
Thanks @nkanazawa1989. It's very good and almost ready to merge. A few small comments:
1. In my opinion it's better to separate files from T1.
2. @coruscating started to add minimum examples to documentation of all the experiments, I guess it means that new experiments should come equipped with an example.
3. For the case of amplitude x-values, the figure x-axis should be quadratically scaled, after #1215 is merged (this is not blocking this PR from merging, this can be done later).

nkanazawa1989 commented on qiskit_experiments/library/characterization/ramsey_xy.py:
```diff
@@ -420,7 +420,7 @@ def parameters(self) -> np.ndarray:
         return opt.delays
 
-    def parameterized_circuits(self) -> Tuple[QuantumCircuit, QuantumCircuit]:
+    def parameterized_circuits(self) -> Tuple[QuantumCircuit, ...]:
```
Just for generalization (I recently added this method to every new experiment requiring parameter scan). In future we can easily add new base class for parameter-scan type experiment with these methods.

nkanazawa1989 commented:
> In my opinion it's better to separate files from T1.

I started to categorize experiments with the type of technologies. `StarkRamsyXY` and `StarkRamseyXYAmpScan` experiment are both implemented in `qiskit_experiments.library.characterization.ramsey_xy` with `RasmeyXY` which is without Stark tone. This is basically a variant of inversion recovery experiment with Stark tone and fixed delay, and I added this experiment to existing t1 module.

> @coruscating started to add minimum examples to documentation of all the experiments, I guess it means that new experiments should come equipped with an example.

I was thinking of this, but this experiment requires full dynamics simulation with calibrated pulses. This needs too lengthy code to fit in with the API doc.

> For the case of amplitude x-values, the figure x-axis should be quadratically scaled, after https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1215 is merged (this is not blocking this PR from merging, this can be done later).

This is good suggestion. I can update after the PR is merged (or I can do this in this PR).

nkanazawa1989 commented:
The `xscale` option is added in [846ca09](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1010/commits/846ca09a094d7a948580c9b222ff508eee0b9a6d). By default the output would look like below.

![image](https://github.com/Qiskit-Extensions/qiskit-experiments/assets/39517270/849072ed-91ce-4dd8-8289-d61102e99da3)

The `xscale` falls back into linear when Stark coefficients are provided.
